weave/trace/vals.py

```diff
@@ -270,8 +300,47 @@ def __eq__(self, other: Any) -> bool:
 
     def _mark_dirty(self) -> None:
         self.table_ref = None
+        self._prefetched_rows = None
         super()._mark_dirty()
 
+    def _local_iter(self) -> Generator[dict, None, None]:
+        """
+        This is the case where we:
+        1. Have all the rows in memory
+        2. Have all the row digests
+
+        In this case, we don't need to make any calls and can just return the rows
+        """
+        if (
+            self.ref is None
+            or self.table_ref is None
+            or self.table_ref.row_digests is None
+            or self._prefetched_rows is None
+        ):
+            logger.error(
+                "Expected all row digests and prefetched rows to be set, falling back to remote iteration"
+            )
+            yield from self._remote_iter()
```

Comment:
i think this should throw instead of falling back, when could this ever occur outside of a programming error? Feels a bit too magical to call "local_iter" and get server results. 

Could also have a parent method `iter` that farms out the iter to either local or remote...

Reply:
I agree. I changed the method to `_local_iter_with_remote_fallback`. The reason i am being careful here is that we are dealing with variations in the backend and i would rather print an error than crash user code in the near term.